parsers/Parser.js

```diff
@@ -7,31 +9,97 @@ module.exports = class Parser {
         this.document = document;
     }
 
-    parse() {
-        return [];
+    getRules() {
+        const rules = this.rules;
+
+        return rules && rules.length ? this.flagDuplicates(rules) : [];
     }
 
-    getRules() {
-        let rules = this.parse();
+    get rules() {
+        // This default implementation fits for for YAML and JSON parsers.
+        // JS & Package parsers override it.
+        return this.getAllRules(this.ast);
+    }
 
-        if (!rules || rules.length === 0) {
-            return rules;
-        }
+    get ast() {
+        throw new Error('ast getter must be implemented in the child class.');
+    }
 
-        return this.flagDuplicates(rules);
+    getNodeKey(node) {
+        // This default implementation fits for YAML, JSON, and Package parsers.
+        // JS parser overrides it.
+        return node.key.value;
+    }
+
+    /** Gets properties of an object-like AST node. */
+    getProps(node) {
+        throw new Error('getProps method must be implemented in the child class.');
+    }
+
+    /** Gets elements of an array-like AST node. */
+    getItems(node) {
+        throw new Error('getItems method must be implemented in the child class.');
+    }
+
+    /** Flatmaps `callback` over `node`'s child property (or children properties) named `key`. */
+    mapChildren(node, key, callback) {
+        return this.getProps(node)
+            .filter(prop => this.getNodeKey(prop) === key)
+            .flatMap(prop => callback(prop.value));
+    }
+
+    /** Gets both plain and overridden rules from `node`. */
+    getAllRules(node) {
+        const mainRules = this.getRulesFromNode(node);
+        const overrideRules = this.getOverridesFromNode(node);
+
+        return mainRules.concat(overrideRules);
+    }
+
+    /** Gets rules from AST node, assuming this node has 'rules' child. */
+    getRulesFromNode(node) {
+        return this.mapChildren(node, 'rules', value =>
+            this.getProps(value).map(rule => this.getRule(rule)),
+        );
+    }
+
+    /** Gets rules from AST node, assuming this node has 'overrides' child. */
+    getOverridesFromNode(node) {
```

Comment:
The function name is misleading.  I read it as `gets all of the overrides from within the given node`, but it is actually `get all of the rules from within all of the overrides from within the given node`.